### PR TITLE
VS-197: Enable wrangler to interact with Vectorize V2 indexes

### PR DESCRIPTION
**Pull request by garvit-gupta**
## What this PR solves / how to test

Enables users to operate on Vectorize V2 operations via wrangler.

Fixes #[VS-197].

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [x] I don't know
  - [ ] Required / Maybe required
  - [ ] Not required because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [] TODO (before merge)
  - [x ] Cloudflare docs PR(s): https://github.com/cloudflare/cloudflare-docs/pull/15916
  - [ ] Not necessary because:

### CODE REVIEW

**Review by ndisidore — APPROVED** [review]
Looks great!

File: packages/wrangler/src/__tests__/vectorize/vectorize.test.ts
Comment (garvit-gupta) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts]: Fixed!

File: packages/wrangler/src/__tests__/vectorize/vectorize.test.ts
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: What does v2 mean when it comes to indexes?
What would happen if test-index was not v2?

**Review by petebacondarwin — COMMENTED** [review]
I was a bit confused by the use of v2 at first.
I think what you are saying is that the v2 flag effectively provides a brand new product from v1, right?
For example it seems that you can create v1 and v2 indexes with the same name and then interact with them separately?
If not then what would it mean to create a v1 index and then use v2 commands to interact with it?

Can you clarify this in the command docs?

File: .changeset/thick-hairs-notice.md
Comment (garvit-gupta) [reply .changeset/thick-hairs-notice.md:2]: Vectorize V1 is an open-beta product, and Vectorize V2 is a re-architectured system that will soon go beta and eventually GA (at which point we will deprecate Vectorize v1).

Vectorize V2 is essentially a different product. V2 operations can only work on a v2 index (created using the use-v2 flag) and the same applies for V1. However, we cannot create a v2 index if a v1 index of the same name exists and vice versa.

File: packages/wrangler/src/__tests__/vectorize/vectorize.test.ts
Comment (garvit-gupta) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: If we try a v2 operation (say get index) on a v1 index, the wrangler client would receive a 403 response from Vectorize. Same goes if we try a v1 operation on a v2 index.

Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: What would be the error message. Will it be clear to the user what went wrong and what to do to fix it?
Also what is the roll off plan for deprecating and removing v1?

Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: For example, for how long will you support `use-v2` as a command line flag?
In general we are moving towards feature flagging things in Wrangler with the `experimental`/`x` prefix. So in this case it would be 

```
wrangler vectorize get test-index --experimental-v2
```
or
```
wrangler vectorize get test-index --x-v2
```

And for stuff you are removing prefixing with `legacy`/`deprecated`. For example:

```
wrangler vectorize get test-index --legacy-v1
```
or
```
wrangler vectorize get test-index --deprecated-v1
```

Comment (garvit-gupta) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: The error message in the apiv4 response format would be like `Index: "${index_name}" with version "${index_version}" called with api version "${operation_version}"` along with a 403 response.

I would let Yevgen address the deprecation plan.

Comment (netgusto) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: cc @sejoker

File: .changeset/thick-hairs-notice.md
Comment (garvit-gupta) [reply .changeset/thick-hairs-notice.md]: Updated the changeset patch to minor, and made the v2 option description in the commands more descriptive. @petebacondarwin I hope that would be sufficient.

**Comment by threepointone** [comment]
Is there a command that converts a v1 index to a v2 index?

**Comment by sejoker** [comment]
> Is there a command that converts a v1 index to a v2 index?

This will be worked on later in the quarter, definitely something we will provide before GA.

File: packages/wrangler/src/__tests__/vectorize/vectorize.test.ts
Comment (sejoker) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: > For example, for how long will you support `use-v2` as a command line flag? In general we are moving towards feature flagging things in Wrangler with the `experimental`/`x` prefix. So in this case it would be
> 
> ```
> wrangler vectorize get test-index --experimental-v2
> ```
> 
> or
> 
> ```
> wrangler vectorize get test-index --x-v2
> ```
> 
> And for stuff you are removing prefixing with `legacy`/`deprecated`. For example:
> 
> ```
> wrangler vectorize get test-index --legacy-v1
> ```
> 
> or
> 
> ```
> wrangler vectorize get test-index --deprecated-v1
> ```

@petebacondarwin the plan to start with private beta this month, once we will be confident to open it up for public beta sometime in August we will stop allowing V1 index creation and V2 only will be used for new indexes.

Comment (sejoker) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts:224]: I like the idea of using `--deprecated-v1` for dealing with existing V1 indexes once V2 goes to public beta.

Comment (garvit-gupta) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts]: We discussed this within our team, and decided to with an `--experimental-v2` flag with a default value of false until we put v1 indexes on the deprecation path. At that time we will switch to `--deprecated-v1`. 

Made the change to update the commands and the related docs.

**Comment by CarmenPopoviciu** [comment]
@garvit-gupta is my understanding correct that some backend work needs to happen before we can merge & release this PR? Are any further wrangler changes required or is the PR in its current state ready to go?

**Comment by CarmenPopoviciu** [comment]
spoke async with the team, and this is now ready for re-review

File: .changeset/thick-hairs-notice.md
Comment (CarmenPopoviciu) [inline .changeset/thick-hairs-notice.md:5]:
```suggestion
feat: Enable Wrangler to operate on Vectorize V2 indexes
```

File: packages/wrangler/src/vectorize/insert.ts
Comment (sejoker) [inline packages/wrangler/src/vectorize/insert.ts:67]: let's put a warning for V2 with batch size > 1 and < 5000:
`The global rate limit for the Cloudflare API of 1200 requests per five minutes. Please use a maximum of 5000 vectors per embeddings.ndjson file to stay within the service limits`

File: .changeset/thick-hairs-notice.md
Comment (garvit-gupta) [reply .changeset/thick-hairs-notice.md]: Updated!

File: packages/wrangler/src/vectorize/insert.ts
Comment (garvit-gupta) [reply packages/wrangler/src/vectorize/insert.ts:67]: Updated!

File: packages/wrangler/src/__tests__/vectorize/vectorize.test.ts
Comment (petebacondarwin) [reply packages/wrangler/src/__tests__/vectorize/vectorize.test.ts]: > We discussed this within our team, and decided to with an `--experimental-v2` flag with a default value of false until we put v1 indexes on the deprecation path. At that time we will switch to `--deprecated-v1`.
> 
> Made the change to update the commands and the related docs.

This is not what I see in the current code for this PR.
I see that there is no --experimental-v2 flag here only the --deprecated-v1 flag, which defaults to false.

I would assume that for the private beta the default needs to switch to true and the --experimental-v2 flag needs to be implemented?

Also, could I call out that it would be much nicer for the user if these commands inferred the version of the index from its type (via a REST request) rather than requiring the user to keep track of which version they are using. The only place one really needs the v1/v2 flag is at index creation time, no?

**Comment by petebacondarwin** [comment]
As I understand it, this PR should only land when we V2 goes into public beta.

File: packages/wrangler/src/vectorize/client.ts
Comment (ndisidore) [inline packages/wrangler/src/vectorize/client.ts:27]: super nit (that's not worth a commit needing a rebuild): in js, camelCase is preferred

**Review by ndisidore — COMMENTED** [review]
This is looking great! I know there's some work you want to do around the query parse function. Happy to approve once we look at that

File: packages/wrangler/src/vectorize/client.ts
Comment (garvit-gupta) [reply packages/wrangler/src/vectorize/client.ts]: Fixed it along with some other changes!

**Comment by garvit-gupta** [comment]
> This is looking great! I know there's some work you want to do around the query parse function. Happy to approve once we look at that

The query filter parse and validate logic looks cleaner now. Ready for review!

File: packages/wrangler/src/vectorize/query.ts
Comment (garvit-gupta) [reply packages/wrangler/src/vectorize/query.ts]: Great catch! Fixed it.

**Comment by sejoker** [comment]
@petebacondarwin @CarmenPopoviciu could you please confirm this PR is ready to be merged from your perspective? Confirmed, we want to release this change once Vectorize V2 goes into public beta.

**Review by petebacondarwin — APPROVED** [review]
What a massive PR! From what I can see this looks great.
Happy for this to be merged when it is unblocked for product release.
cc @RamIdeas - you might be interested in this contribution.

File: packages/wrangler/src/vectorize/query.ts
Comment (garvit-gupta) [reply packages/wrangler/src/vectorize/query.ts:27]: The `query` operation was never enabled for Vectorize V1 in Wrangler (if you check index.ts, you'll see that it was enabled as a part of this PR). So it is not a breaking change. `--vector` makes this param consistent with the ApiV4 route params and would be the preferable option.

File: packages/wrangler/src/vectorize/common.ts
Comment (garvit-gupta) [reply packages/wrangler/src/vectorize/common.ts]: That's a good idea. Included this change with the latest update.

**Review by ndisidore — APPROVED** [review]

**Review by sejoker — APPROVED** [review]

**Comment by sejoker** [comment]
@RamIdeas @petebacondarwin we are happy to merge the change and release it next week.

**Pull request merged**
